### DISCUSSION TITLE
How to update formatting of inline code block with tailwind css?

### DISCUSSION THREAD

**Discussion by ian-whitestone** [markdoc/markdoc#205] [Q&A] [answered]
Hey folks!

I am currently building a blog using a [TailwindUI](https://tailwindui.com/) template called [syntax](https://syntax.tailwindui.com/), which leverages markdoc under the hood. The template uses React, Next JS and tailwind CSS.

Right now if I have some markdown that looks like this:

```
Let’s extend the example, by zooming out to several micro partitions and their min and max values for the `created_at` column.

{% blogPostImage imageName="pruning" width="2/3" /%}

If a user were to run the following query:

``sql
SELECT *
FROM orders
WHERE
    created_at > DATE'2022/09/12'
``

^^ these are actually triple backticks

```

The inline code block will get rendered with the backticks (you can view this on the [live website](https://select.dev/posts/snowflake-micro-partitions#pruning) if you prefer).

![CleanShot 2022-09-20 at 12 09 45@2x](https://user-images.githubusercontent.com/11096727/191309711-87aa6abe-d861-440f-b365-b615302f6871.png)

The other code block renders fine, and it is being formatted based on the [syntax highlighting example](https://markdoc.dev/docs/examples#syntax-highlighting) from your docs.

I am wondering how I could change the formatting of this? I'd like it to look more like what you guys have on your site:

![image](https://user-images.githubusercontent.com/11096727/191310196-112a4350-b495-47df-81dc-a0eae88daa28.png)

Here's what the current set up is for `tags.js` and `nodes.js` if it helps.

<details>

<summary> tags.js </summary>

```js
import { Callout } from '@/components/Callout'
import { QuickLink, QuickLinks } from '@/components/QuickLinks'

import { blogPostImage } from '@/components/BlogImages'

const tags = {
  callout: {
    attributes: {
      title: { type: String },
      type: {
        type: String,
        default: 'note',
        matches: ['note', 'warning'],
        errorLevel: 'critical',
      },
    },
    render: Callout,
  },
  figure: {
    selfClosing: true,
    attributes: {
      src: { type: String },
      alt: { type: String },
      caption: { type: String },
    },
    render: ({ src, alt = '', caption }) => (
      <figure>
        {/* eslint-disable-next-line @next/next/no-img-element */}
        <img src={src} alt={alt} />
        <figcaption>{caption}</figcaption>
      </figure>
    ),
  },
  'quick-links': {
    render: QuickLinks,
  },
  'quick-link': {
    selfClosing: true,
    render: QuickLink,
    attributes: {
      title: { type: String },
      description: { type: String },
      icon: { type: String },
      href: { type: String },
    },
  },
  blogPostImage: {
    selfClosing: true,
    render: blogPostImage,
    attributes: {
      imageName: { type: String },
      width: { type: String },
    },
  },
}

export default tags
```

</details>

<details>

<summary> nodes.js </summary>

```js
import { Fence } from '@/components/Fence'
import { nodes as defaultNodes } from '@markdoc/markdoc'

const nodes = {
  document: {
    render: undefined,
  },
  th: {
    ...defaultNodes.th,
    attributes: {
      ...defaultNodes.th.attributes,
      scope: {
        type: String,
        default: 'col',
      },
    },
  },
  fence: {
    render: Fence,
    attributes: {
      language: {
        type: String,
      },
    },
  },
}

export default nodes
```

</details>

**Comment by mfix-stripe** [comment, marked as answer]
Hey @ian-whitestone, the backticks are part of Tailwind's Syntax CSS itself. To remove the backticks, you can do:
```css
.prose :where(code):not(:where([class~=not-prose] *)):before,
.prose :where(code):not(:where([class~=not-prose] *)):after {
  content: ""
}
```

which is just undoing the styles set by Syntax, as you can see here:

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/62121649/191312068-84511893-ff99-4bef-8bd3-97fd7ce110bb.png">

As for how we style our inline `code` elements on markdoc.dev, [here is the CSS](https://github.com/markdoc/docs/blob/main/public/globals.css#L283-L293): 
```css
code {
  font-size: inherit;
  font-size: 15px;
  font-family: var(--mono);
  font-weight: 500;
  padding: 1px 2px;
  background: var(--code-background);
  border: 1px solid var(--code-border);
  border-radius: 3px;
  white-space: nowrap;
}
```

**Comment by ian-whitestone** [comment]
Thanks @mfix-stripe , really appreciate the quick response!

That seems like a good path forward for removing the backticks, but i would ultimately like to apply some custom CSS styling for inline code blocks like you all do on your site.

Would it be possible to override how markdoc is rendering backticks by providing a custom node? Similar to what is done in the [example provided here](https://markdoc.dev/docs/examples#syntax-highlighting)?


Here's what the Syntax template I'm using is currently doing for full code blocks:

```js
// nodes.js
import { Fence } from '@/components/Fence'
import { nodes as defaultNodes } from '@markdoc/markdoc'

const nodes = {
  document: {
    render: undefined,
  },
  th: {
    ...defaultNodes.th,
    attributes: {
      ...defaultNodes.th.attributes,
      scope: {
        type: String,
        default: 'col',
      },
    },
  },
  fence: {
    render: Fence,
    attributes: {
      language: {
        type: String,
      },
    },
  },
}

export default nodes
```

```jsx
// components/Fence.jsx
import { Fragment } from 'react'
import Highlight, { defaultProps } from 'prism-react-renderer'

export function Fence({ children, language }) {
  return (
    <Highlight
      {...defaultProps}
      code={children.trimEnd()}
      language={language}
      theme={undefined}
    >
      {({ className, style, tokens, getTokenProps }) => (
        <pre className={className} style={style}>
          <code>
            {tokens.map((line, lineIndex) => (
              <Fragment key={lineIndex}>
                {line
                  .filter((token) => !token.empty)
                  .map((token, tokenIndex) => (
                    <span key={tokenIndex} {...getTokenProps({ token })} />
                  ))}
                {'\n'}
              </Fragment>
            ))}
          </code>
        </pre>
      )}
    </Highlight>
  )
}
```


Some I'm wondering if I could add a similar custom component to render the `inline code block` in a better way, I'm imagining something like this:

```js
// Components/InlineCode.jsx

export function InlineCode({ children }) {
  return (
    <pre className="font-mono bg-slate-700 border-2 font-medium">
        <code>{children}</code>
    </pre>
  )
}
```

Apologies for the potentially naive question, I have 0 idea what I'm doing 😄

**Reply by mfix-stripe** [reply]
@ian-whitestone yes, you can add a custom `code` node just like you have done for `fence`, but use `render: InlineCode` instead.

This said, you don't need to, since this is all just CSS as far as I can tell, so you can just use the CSS styles I included at the bottom of my previous message. 

Up to you 🙂

**Reply by ian-whitestone** [reply]
Thanks! Yeah I may try that first, I just wasn't sure whether it would interfere with the existing styles being used for the main code blocks (with ```), since i want those to stay as is.

**Reply by mfix-stripe** [reply]
Yeah makes sense! Styling and Markdoc are separate concern, but they do overlap sometimes, depending on your implementation. 

Even if you override the `code` node, you might still see the backticks since it looks like Tailwind applies them to any `<code>` element. So you might still need to add the style overrides I shared above. 

I don't expect these to conflict with your `fence` (```) styles, though.

**Reply by ian-whitestone** [reply]
For others reference, here's what I ended up doing:

```js
// new component file
export function InlineCode({ content }) {
    return (
        <span className="bg-slate-200 dark:bg-slate-700 px-0.5 rounded ">
            <code className="font-mono font-medium text-sky-500 dark:text-sky-300">{content}</code>
        </span>
    )
}
```


```js
// nodes.js
const nodes = {
  // ...
  code: {
    render: InlineCode,
    attributes: {
      content: { type: String },
    }
  }
}

export default nodes
```

and in `tailwind.config.js` you can get the backticks removed by following the solution here: https://github.com/tailwindlabs/tailwindcss-typography/issues/18

### results

dark mode:
![image](https://user-images.githubusercontent.com/11096727/191556550-dd7cee20-08e2-483b-837a-5a38e56b731e.png)

light mode
![image](https://user-images.githubusercontent.com/11096727/191556585-fa918151-b729-4991-8570-ea6b30f5b869.png)


i'm still fiddling with the colouring, font-weight and background contrast, but you get the gist of it!

**Reply by mfix-stripe** [reply]
It looks fantastic @ian-whitestone! 🙌 Nice work

**Comment by sasha-computer** [comment]
Hey @mfix-stripe, I'm asking here because I have the exact same issue with nextJS/markdoc. You mentioned this code snippet:

```
.prose :where(code):not(:where([class~=not-prose] *)):before,
.prose :where(code):not(:where([class~=not-prose] *)):after {
  content: ""
}
```

My situation is a little more complicated as I have a custom code component that already does a lot, but for all inline code blocks, I want to remove the backticks and add css. How would I achieve this? Where do I put the code snippet you shared? 
`tailwind.config.js`? 

Any help would be greatly appreciated. I have read through the tailwind issue above and that seems to give a solution for each inline code block as oppose to the inline code blocks rendered by markdoc. I have also read the tailwind docs, and I'm still stumped.

Thanks

**Reply by mfix-stripe** [reply]
You can add your custom CSS to a global CSS file in your Next.js app (nothing to do with Tailwind): https://nextjs.org/learn/basics/assets-metadata-css/global-styles

**Reply by sasha-computer** [reply]
Thanks so much, really appreciate the help. That worked!
